feat: created alert style popover variations (#4058)

agagancarczyk opened the pull request:
![Screenshot 2021-05-11 at 10 36 54](https://user-images.githubusercontent.com/4890675/117800317-ee48bb00-b24a-11eb-8245-1582da7b44aa.png)
![Screenshot 2021-05-11 at 11 16 00](https://user-images.githubusercontent.com/4890675/117800337-f143ab80-b24a-11eb-8af4-c4bd097904bd.png)

patternfly-build commented:
Preview: https://patternfly-pr-4060.surge.sh

agagancarczyk commented:
hi @mcoker this PR is ready for review, but it looks like all test_a11 are failing. Is there something I need to do to make them all pass? Thanks

mcoker commented:
Hi @agagancarczyk thanks for the contribution! A popover is very similar to a modal, and we recently introduced this same feature into the modal component, so for consistency it would be good to mimic that implementation. You can see that in the component at https://www.patternfly.org/v4/components/modal/html#icon and in https://github.com/patternfly/patternfly/blob/master/src/patternfly/components/ModalBox/modal-box.scss

I will leave more specific comments in the PR code!

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
````diff
@@ -132,34 +140,126 @@ import './Popover.css'
 {{/popover}}
 ```
 
+### Variations
````
Can you create unique example names for these variations? A default "Icon" example with "Success alert", "Warning alert", etc.

agagancarczyk replied:
This is fixed now. Please advise if the names are fine. I can still change them.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-success" title--attribute='id="popover-top-header"'}}
+      Success Popover header
+    {{/title}}
+    {{#> popover-body popover-body--attribute='id="popover-top-body"'}}
```
Since this generates an `id` attribute on this element, it needs to be unique on the page, and there is already an example with the name "popover-top-body" on this page, which is generating the a11y error in the PR build.

```suggestion
    {{#> popover-body popover-body--attribute='id="popover-success-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{/button}}
+    {{#> popover-icon popover-icon--type="check-circle"}}
+    {{/popover-icon}}
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-success" title--attribute='id="popover-top-header"'}}
```
```suggestion
    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-success" title--attribute='id="popover-success-alert-header"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
````diff
@@ -132,34 +140,126 @@ import './Popover.css'
+### Variations
+
+```hbs
+{{#> popover popover--modifier="pf-m-top pf-m-success" popover--attribute='aria-labelledby="popover-top-header" aria-describedby="popover-top-body"'}}
````
```suggestion
{{#> popover popover--modifier="pf-m-top pf-m-success" popover--attribute='aria-labelledby="popover-success-alert-header" aria-describedby="popover-success-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+  {{/popover-content}}
+{{/popover}}
+<br/>
+{{#> popover popover--modifier="pf-m-top pf-m-error" popover--attribute='aria-labelledby="popover-top-header" aria-describedby="popover-top-body"'}}
```
```suggestion
{{#> popover popover--modifier="pf-m-top pf-m-error" popover--attribute='aria-labelledby="popover-error-alert-header" aria-describedby="popover-error-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{/button}}
+    {{#> popover-icon popover-icon--type="exclamation-circle"}}
+    {{/popover-icon}}
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-error" title--attribute='id="popover-top-header"'}}
```
```suggestion
    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-error" title--attribute='id="popover-error-alert-header"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-error" title--attribute='id="popover-top-header"'}}
+      Error Popover header
+    {{/title}}
+    {{#> popover-body popover-body--attribute='id="popover-top-body"'}}
```
```suggestion
    {{#> popover-body popover-body--attribute='id="popover-error-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+  {{/popover-content}}
+{{/popover}}
+<br/>
+{{#> popover popover--modifier="pf-m-top pf-m-warning" popover--attribute='aria-labelledby="popover-top-header" aria-describedby="popover-top-body"'}}
```
```suggestion
{{#> popover popover--modifier="pf-m-top pf-m-warning" popover--attribute='aria-labelledby="popover-warning-alert-header" aria-describedby="popover-warning-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{/button}}
+    {{#> popover-icon popover-icon--type="exclamation-triangle"}}
+    {{/popover-icon}}
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-warning" title--attribute='id="popover-top-header"'}}
```
```suggestion
    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-warning" title--attribute='id="popover-warning-alert-header"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-warning" title--attribute='id="popover-top-header"'}}
+      Warning Popover header
+    {{/title}}
+    {{#> popover-body popover-body--attribute='id="popover-top-body"'}}
```
```suggestion
    {{#> popover-body popover-body--attribute='id="popover-warning-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+  {{/popover-content}}
+{{/popover}}
+<br/>
+{{#> popover popover--modifier="pf-m-top pf-m-info" popover--attribute='aria-labelledby="popover-top-header" aria-describedby="popover-top-body"'}}
```
```suggestion
{{#> popover popover--modifier="pf-m-top pf-m-info" popover--attribute='aria-labelledby="popover-info-alert-header" aria-describedby="popover-info-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{/button}}
+    {{#> popover-icon popover-icon--type="info-circle"}}
+    {{/popover-icon}}
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-info" title--attribute='id="popover-top-header"'}}
```
```suggestion
    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-info" title--attribute='id="popover-info-alert-header"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
+    {{#> title titleType="h1" title--modifier="pf-m-md pf-m-info" title--attribute='id="popover-top-header"'}}
+      Info Popover header
+    {{/title}}
+    {{#> popover-body popover-body--attribute='id="popover-top-body"'}}
```
```suggestion
    {{#> popover-body popover-body--attribute='id="popover-info-alert-body"'}}
```

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -1,8 +1,14 @@
-  --pf-c-popover--MinWidth: calc(var(--pf-c-popover__content--PaddingLeft) + var(--pf-c-popover__content--PaddingRight) + #{pf-size-prem(300px)});
-  --pf-c-popover--MaxWidth: calc(var(--pf-c-popover__content--PaddingLeft) + var(--pf-c-popover__content--PaddingRight) + #{pf-size-prem(300px)});
+  --pf-c-popover--MinWidth:
+    calc(
```
We generally try and keep the CSS functions on a single line when possible, unless it's very long and hard to read. Would you mind making this a single line and putting it inline with `--pf-c-popover--MinWidth`?

mcoker replied:
This applies to the entire stylesheet, so I won't comment on other instances.

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
       left: 50%;
-      transform: translateX(var(--pf-c-popover__arrow--m-top--TranslateX)) translateY(var(--pf-c-popover__arrow--m-top--TranslateY)) rotate(var(--pf-c-popover__arrow--m-top--Rotate));
+      transform:
+        translateX(var(--pf-c-popover__arrow--m-top--TranslateX))
```
We typically keep properties with multiple values on a single line, though if we do break them up into multiple lines, they're all on new lines. In this case, I would just put the `transform` functions on a single line.

mcoker replied:
This applies to other instances in the stylesheet, so I won't comment on those.

agagancarczyk replied:
This is fixed now.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+  &.pf-m-success {
+    --pf-c-popover__icon--Color: var(--pf-c-popover--m-success__icon--Color);
+
+    .pf-c-title {
```
If you use the modal component as an example, I think a good way to organize this would be to add a `.pf-m-icon` class to the `.pf-c-popover` element, which sets `display: flex;` on `.pf-c-title`. That provides an opt-in to the display change on the title to keep from breaking existing users, and it allows us to place a `<span class="pf-c-popover__title-icon">` element in the title for the icon and a `<span class="pf-c-popover__title-text">` in the title for the text.

Then instead of styling `.pf-c-title` inside of each `pf-m-[alert status]` modifier, you can set `margin-left` on `.pf-c-popover__title-text` once.

You can then also introduce a variable to set the text color (`--pf-c-popover--m-success__title-text--Color`) to control the color instead of needing to introduce variations on the title component itself, to keep the changes localized to the popover component.

agagancarczyk replied:
Thank you! This is fixed now. I used  modalBox component as an example and reworked the popover variations.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+  &.pf-m-success {
+    --pf-c-popover__icon--Color: var(--pf-c-popover--m-success__icon--Color);
+
+    .pf-c-title {
```
If you use the modal component as an example, I think a good way to organize this would be to add a `.pf-m-icon` class to the `.pf-c-popover` element, which sets `display: flex;` on `.pf-c-title`. That provides an opt-in to the display change on the title to keep from breaking existing users, and it allows us to place a `<span class="pf-c-popover__title-icon">` element in the title for the icon and a `<span class="pf-c-popover__title-text">` in the title for the text.

Then instead of styling `.pf-c-title` inside of each `pf-m-[alert status]` modifier, you can set `margin-left` on `.pf-c-popover__title-text` once.

You can then also introduce a variable to set the text color (`--pf-c-popover--m-success__title-text--Color`) to control the color instead of needing to introduce variations on the title component itself, to keep the changes localized to the popover component.

agagancarczyk replied:
Thank you! This is fixed now. I used  modalBox component as an example and reworked the popover variations.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+  }
+
+  .pf-c-popover__icon {
+    position: absolute;
```
using `position: absolute` doesn't scale very well since it takes the element out of the document flow. For example, if a user introduces a larger image here than the one we've accounted for, they will also need to update the title's `margin-left` to account for the image size/width.

You should just be able to use regular positioning, and remove this and the `right` properties if you move the icon in the title, and make the title a flex layout.

agagancarczyk replied:
This is not applicable anymore as I used modalBox component to reimplement the popover.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+    right: var(--pf-c-popover--c-button--Left);
+    display: flex;
+    margin-top: var(--pf-c-popover__icon--MarginTop);
+    font-size: var(--pf-c-popover__icon--FontSize);
```
If you move the icon in the title, you can remove `font-size` since it will inherit the title font size.

agagancarczyk replied:
This is not applicable anymore as I used modalBox component to reimplement the popover.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+    right: var(--pf-c-popover--c-button--Left);
+    display: flex;
+    margin-top: var(--pf-c-popover__icon--MarginTop);
+    font-size: var(--pf-c-popover__icon--FontSize);
```
If you move the icon in the title, you can remove `font-size` since it will inherit the title font size.

agagancarczyk replied:
This is not applicable anymore as I used modalBox component to reimplement the popover.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+    position: absolute;
+    right: var(--pf-c-popover--c-button--Left);
+    display: flex;
+    margin-top: var(--pf-c-popover__icon--MarginTop);
```
You shouldn't need this if you move the icon in the title.

agagancarczyk replied:
This is not applicable anymore as I used modalBox component to reimplement the popover.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -67,38 +94,90 @@
+  .pf-c-popover__icon {
+    position: absolute;
+    right: var(--pf-c-popover--c-button--Left);
+    display: flex;
```
You shouldn't need to set this.

agagancarczyk replied:
This is not applicable anymore as I used modalBox component to reimplement the popover.

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -132,34 +140,126 @@ import './Popover.css'
-| `.pf-m-no-padding` | `.pf-c-popover` | Removes the outer padding from the popover content. |
-| `.pf-m-width-auto` | `.pf-c-popover` | Allows popover width to be defined by the popover content. |
+
+| Class                    | Applied to                                | Outcome                                                                                                                                   |
```
We used to format our markdown tables this way but found the maintenance to maintain the whitespace more work than it's worth, and just use single spaces in the tables now. You can see that across the library in our other examples. Would you mind being consistent with the markdown tables in the rest of the component documentation

agagancarczyk replied:
This is fixed now. I had a setting in VSCode which was formatting the code.

mcoker commented on src/patternfly/components/Popover/popover-icon.hbs:
```diff
@@ -0,0 +1,6 @@
+  {{#if popover-icon--attribute}}
+    {{{popover-icon--attribute}}}
+  {{/if}}>
+  <i class="fas fa-fw{{#if popover-icon--type}} fa-{{popover-icon--type}}{{/if}}" aria-hidden="true"></i>
```
Nice touch using `popover-icon--type`!!

agagancarczyk replied:
:)

mcoker reviewed (changes requested):
Thanks for the contribution, and great work! Left some comments for review. Also would you mind removing `package-lock.json` from the file changeset? We don't commit that with PRs.

mcoker commented:
Hi @agagancarczyk just following up to see if you had a chance to review the PR comments and if you had any questions.

agagancarczyk commented:
hi @mcoker many thanks for the feedback. I hope that I have addressed all issues. As suggested I used modalBox component as an example here. Please let me know if you think there is more for fixing. This is the final result. 
<img width="235" alt="Screenshot 2021-05-28 at 17 56 40" src="https://user-images.githubusercontent.com/4890675/120022753-c0060200-bfe4-11eb-94a6-ccbec0ca1c0c.png">

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
````diff
@@ -131,35 +138,159 @@ import './Popover.css'
+{{/popover}}
+```
+### Success alert
+
````
Nit - can you remove the space between the header and the start of the hbs block backticks?

agagancarczyk replied:
Fixed

mcoker commented on src/patternfly/components/Popover/examples/Popover.md:
````diff
@@ -131,35 +138,159 @@ import './Popover.css'
+    {{/popover-footer}}
+  {{/popover-content}}
+{{/popover}}
+```
````
Nit - can you add a space between the end of a block and the header starting the next block?

agagancarczyk replied:
Fixed

mcoker commented on src/patternfly/components/Title/title.scss:
```diff
@@ -31,6 +31,18 @@
   --pf-c-title--m-md--FontSize: var(--pf-global--FontSize--md);
```
I don't think we'll need these variations on the title component - we should just be able to keep the status colors and treatment in the popover.

agagancarczyk replied:
hi @mcoker would you like me to remove all code I added for title variations (also from Title/examples/Title.md)?

mcoker replied:
@agagancarczyk yep, we shouldn't need any changes to the title component to support this feature.

agagancarczyk replied:
thank you @mcoker , this is fixed now.

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -94,11 +129,78 @@
 .pf-c-popover__content {
   position: relative;
-  padding: var(--pf-c-popover__content--PaddingTop) var(--pf-c-popover__content--PaddingRight) var(--pf-c-popover__content--PaddingBottom) var(--pf-c-popover__content--PaddingLeft);
+  padding:
```
Can you move these values so everything is on a single line?

agagancarczyk replied:
Fixed

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -94,11 +129,78 @@
+  @include pf-text-overflow;
+}
+
+.pf-c-popover__title {
```
I created an issue to add a `__title` and `__header` element to the popover to match other components (like the modal) here - https://github.com/patternfly/patternfly/issues/4123

I don't think we need the `__title` now, as we'll need to maintain 2 separate elements - this new `__title` and the existing `.pf-c-title` to avoid a breaking change. I think we can do everything we need for these alert variations working within the existing title.

agagancarczyk replied:
sure @mcoker , I have now removed all occurrences of .pf-c-popover__title from popover.scss

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -94,11 +129,78 @@
   }
+
+  &.pf-m-icon {
+    display: flex;
```
I'm not sure if you need this to change the layout of the popover component itself. This class could be used to target the title via `.pf-c-popover__content > .pf-c-title { display: flex; }` though.

agagancarczyk replied:
Fixed

mcoker commented on src/patternfly/components/Popover/popover.scss:
```diff
@@ -5,13 +5,48 @@
+  --pf-c-popover__title--FontSize: var(--pf-global--FontSize--md);
+
+  // Title Default
+  --pf-c-title--m-default__title--Color: var(--pf-global--default-color--200);
```
I don't think you'll need these variations on the title, though fwiw we only keep the parent component vars here, so we would have these in the title component instead of the popover component.

agagancarczyk replied:
Thank you @mcoker, will this also be addressed in #4123? if I remove the variations then the icons for popover variations will lose the colours as below:

![after](https://user-images.githubusercontent.com/4890675/125684181-54b66968-9436-445a-8964-c72e881b7ba3.png)

mcoker reviewed (changes requested):
Thanks for making updates! I left some more feedback for review.

Also the build is failing - can you remove `package-lock.json` from the commit? I'm wondering if that is the problem.

mcoker commented:
Hey @agagancarczyk apologies for the delay! I just wanted to check in and see if you'd like to continue working on this PR and if there is anything I can do to help? I'm happy to answer any questions, pair program with you if possible, or either push changes to this PR or issue a PR to your branch for review if you'd like a hand with any of it.

agagancarczyk commented:
hi Michael, I will try to finalize the work over the weekend. Apologies for
the delay.

On Thu, Jul 8, 2021 at 8:40 PM Michael Coker ***@***.***>
wrote:

> Hey @agagancarczyk <https://github.com/agagancarczyk> apologies for the
> delay! I just wanted to check in and see if you'd like to continue working
> on this PR and if there is anything I can do to help? I'm happy to answer
> any questions, pair program with you if possible, or either push changes to
> this PR or issue a PR to your branch for review if you'd like a hand with
> any of it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/patternfly/patternfly/pull/4060#issuecomment-876695294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFKAM6PQOM5UHHOKXX4C53TWX5JLANCNFSM44VAVW7Q>
> .
>


-- 
*Kind Regards,*
*Agnieszka*

mcoker commented:
Hey @agagancarczyk, we're hoping to land this feature in the upcoming sprint with a code freeze EOD 7/28. Do you think we can complete this work by then?

agagancarczyk commented on src/patternfly/components/Popover/examples/Popover.md:
```diff
@@ -131,35 +138,159 @@ import './Popover.css'
 | -- | -- | -- |
-| `role="dialog"` | `.pf-c-popover` | Identifies the element that serves as the popover container. **Required**|
-| `aria-labelledby="[id value of .pf-c-title]"` | `.pf-c-popover` | Gives the popover an accessible name by referring to the element that provides the dialog title. **Required when .pf-c-title is present** |
+| `role="dialog"` | `.pf-c-popover` | Identifies the element that serves as the popover container **Required** |
```
Fixed.

agagancarczyk commented on src/patternfly/components/Popover/popover-title-icon.hbs:
```diff
@@ -0,0 +1,18 @@
+  {{#if popover-title-icon--attribute}}
+    {{{popover-title-icon--attribute}}}
+  {{/if}}>
+  {{#if popover--IsDefaultAlert}}
```
This is great. I'll use ~

agagancarczyk commented:
hi @mcoker I have made the requested changes except for removing colour styles for the icons in the variations popups (src/patternfly/components/Popover/popover.scss). This is because the removal will result in unstyled icons and I'm not sure if this will be covered in a new #4123. Please let me know if you still want me to remove these styles and could you please review the changes? Many thanks!

mcoker reviewed (changes requested):
With the `package-lock.json`, sorry I probably didn't explain it well. We leave the file in the repo, just don't commit changes to it unless we're also making `package.json` changes. I restored that file, so you'll want to pull down that change. This looks great, though I have a bit of feedback below. Let me know if you have questions!

mcoker commented:
Looks like this PR is stale - we have this issue on our roadmap and can issue a new PR.
